### ISSUE TITLE
[ascend] Compile abort: ConvertLinalgRToBinary fails with 'The buffer memory has been released' on a kernel mixing loop-carried int64 vectors and 2D broadcast stores

### ISSUE THREAD

**Issue by jianzs** [triton-lang/triton-ascend#2353]
## Summary

A vLLM inference-time hash kernel fails to compile on triton-ascend with a backend abort
at the `ConvertLinalgRToBinary` stage:

```
triton.compiler.errors.MLIRCompilationError:
[ConvertLinalgRToBinary] encounters error:
LLVM ERROR: The buffer memory has been released and cannot be used again!
```

The kernel is self-contained in the reproducer below (~90 lines + launcher, N = 16 tokens
on one NPU, no vLLM or external data needed). The failure happens deterministically on
two independent machines.

We bisected the kernel into variants sharing the same inputs. The failure requires the
full body ("runtime `while` binary search + full chunk processing + 2D broadcast
arithmetic"); each of the removed pieces alone is fine:

| variant | content | compile outcome |
|---|---|---|
| V0 (full) | `while` + all following code | **FAIL (the error above)** |
| V1 | drop the `while` loop only (keep everything below, incl. 2D broadcast/store) | OK |
| V7 | `while` only + store `req` (drop chunk processing) | OK |
| V8 | `while` + full chunk processing, but **disable `if shift > 0:** (broadcast section)** | OK |
| V13 | `while` + the `prime`/`offset` loads + 1D sums (no broadcast, no 2D store) | OK |
| V15 | `while` + the 2D masked store, with broadcast constants only | OK |
| V14 | `while` + broadcast arithmetic (`rolling[:, None] % prime[None, :] + offset[None, :]`), no 2D store | **FAIL** |
| V10 | like V14 but replace `%` with `+` (plain 2D broadcast add) | **FAIL** |

i.e. neither the `while` loop nor the broadcast/2D-store section alone breaks; removing
the `%` is not enough either (plain broadcast add also aborts). The crash needs the full
"loop-carried int64 vectors feed a long chunk-processing body that ends in a 2D broadcast
store path" composition. The specific `LLVM ERROR: The buffer memory has been released and
cannot be used again!` suggests a backend alloc/lifetime bug (a buffer being re-released
in `ConvertLinalgRToBinary`).

## Reproducer

Attached as `kernel_full_repro.py` (see the issue body text below; runs with
`python3 kernel_full_repro.py` on any Ascend NPU with torch_npu + triton-ascend installed).
The kernel is designed so it does not need the original model, tokenizer, weights or any
external file.

```python
import torch, torch_npu, triton, triton.language as tl

@triton.jit
def _hash_ids_kernel(
    input_ids, token_map, dead_mask, positions, block_table, query_start_loc,
    multipliers, primes, offsets, cache, lookback_token_ids,
    lookback_dead_mask, output,
    num_tokens, num_slots, pad_id, input_stride, mask_stride, position_stride,
    table_stride, table_col_stride, query_stride, num_query_rows, num_table_rows,
    max_blocks, cache_block_size,
    MAX_NGRAM: tl.constexpr, num_heads, BLOCK_T: tl.constexpr,
    BLOCK_H: tl.constexpr, dead_id, lookback_depth, lookback_row_stride,
    lookback_col_stride, lookback_mask_row_stride, lookback_mask_col_stride,
):
    token = tl.program_id(0) * BLOCK_T + tl.arange(0, BLOCK_T)
    layer = tl.program_id(1)
    num_layers = tl.num_programs(1)
    valid = token < num_tokens

    lo = tl.full((BLOCK_T,), 0, tl.int32)
    hi = tl.full((BLOCK_T,), num_query_rows, tl.int32)
    while tl.sum((lo < hi).to(tl.int32), 0) > 0:
        mid = (lo + hi) // 2
        end = tl.load(
            query_start_loc + (mid + 1) * query_stride, lo < hi, other=0,
        )
        right = token >= end
        active = lo < hi
        lo = tl.where(active & right, mid + 1, lo)
        hi = tl.where(active & ~right, mid, hi)
    req = tl.minimum(lo, num_query_rows - 1).to(tl.int64)

    chunk_idx = tl.load(query_start_loc + req * query_stride)
    chunk_idx = tl.minimum(chunk_idx, num_tokens - 1).to(tl.int64)
    chunk_start = tl.load(positions + chunk_idx * position_stride)
    position = tl.load(positions + token * position_stride, valid, other=0).to(tl.int64)
    head = tl.arange(0, BLOCK_H)
    blocked = tl.full((BLOCK_T,), False, tl.int1)
    rolling = tl.full((BLOCK_T,), 0, tl.int64)

    for shift in tl.static_range(MAX_NGRAM):
        lookback = position - shift
        in_batch = lookback >= chunk_start
        batch_idx = tl.maximum(token - shift, 0)
        batch_token = tl.load(input_ids + batch_idx * input_stride, valid & in_batch, other=0)
        batch_source = tl.load(token_map + batch_token, valid & in_batch, other=0)
        batch_dead = tl.load(dead_mask + batch_idx * mask_stride, valid & in_batch, other=False)
        batch_source = tl.where(batch_dead, dead_id, batch_source)

        col = chunk_start - 1 - lookback
        in_window = valid & ~in_batch & (col >= 0) & (col < lookback_depth)
        col = tl.minimum(tl.maximum(col, 0), lookback_depth - 1)
        window_token = tl.load(
            lookback_token_ids + req * lookback_row_stride + col * lookback_col_stride,
            in_window, other=-1,
        )
        known = in_window & (window_token >= 0)
        window_source = tl.load(token_map + window_token, known, other=0)
        window_dead = tl.load(
            lookback_dead_mask
            + req * lookback_mask_row_stride
            + col * lookback_mask_col_stride,
            known, other=False,
        )
        window_source = tl.where(window_dead, dead_id, window_source)

        if cache is not None:
            clamped = tl.minimum(tl.maximum(lookback, 0), max_blocks * cache_block_size - 1)
            block_row = tl.minimum(req, num_table_rows - 1)
            needs_cache = valid & ~in_batch & ~known
            block = tl.load(
                block_table
                + block_row * table_stride
                + (clamped // cache_block_size) * table_col_stride,
                needs_cache, other=0,
            ).to(tl.int64)
            slot = tl.minimum(
                tl.maximum(block * cache_block_size + clamped % cache_block_size, 0),
                num_slots - 1,
            )
            fallback = tl.load(cache + slot, needs_cache, other=0)
        else:
            fallback = tl.full((BLOCK_T,), pad_id, tl.int32)

        source = tl.where(
            in_batch, batch_source, tl.where(known, window_source, fallback)
        ).to(tl.int64)

        blocked |= (lookback < 0) | (source == dead_id)
        value = tl.where(blocked, pad_id, source)
        multiplier = tl.load(multipliers + layer * MAX_NGRAM + shift)
        rolling ^= value * multiplier

        if shift > 0:
            col2 = (shift - 1) * num_heads + head
            param_offset = layer * (MAX_NGRAM - 1) * num_heads + col2
            prime = tl.load(primes + param_offset, head < num_heads, other=1)
            offset = tl.load(offsets + param_offset, head < num_heads, other=0)
            hashed = rolling[:, None] % prime[None, :] + offset[None, :]
            out_offset = (token.to(tl.int64) * num_layers + layer)[:, None] * (
                (MAX_NGRAM - 1) * num_heads
            ) + col2[None, :]
            tl.store(output + out_offset, hashed, valid[:, None] & (head < num_heads))


def main():
    dev = torch.device("npu:0")
    T, L, H, NG = 16, 2, 4, 3
    ipt = torch.randint(0, 100, (T,), device=dev)
    tmap = torch.randint(0, 100, (128,), device=dev)
    dead = torch.zeros(T, dtype=torch.bool, device=dev)
    pos = torch.arange(T, device=dev) + 4
    bt = torch.zeros(2, 4, dtype=torch.int32, device=dev)
    qsl = torch.tensor([0, 8, 16], device=dev, dtype=torch.int32)
    mul = torch.ones(L * NG, dtype=torch.int64, device=dev)
    pr = torch.full((L * (NG - 1) * H,), 97, dtype=torch.int64, device=dev)
    off = torch.zeros(L * (NG - 1) * H, dtype=torch.int64, device=dev)
    cch = torch.zeros(4 * 16, dtype=torch.int32, device=dev)
    lbt = torch.zeros(2, 2, dtype=torch.int32, device=dev)
    lbm = torch.zeros(2, 2, dtype=torch.bool, device=dev)
    out = torch.zeros(T * L * (NG - 1) * H + 1024, dtype=torch.int32, device=dev)
    _hash_ids_kernel[(triton.cdiv(T, 32), L)](
        ipt, tmap, dead, pos, bt, qsl, mul, pr, off, cch, lbt, lbm, out,
        T, cch.shape[0], 0, 1, 1, 1, 4, 1, 1, 1, 4, 16, 16,
        MAX_NGRAM=NG, num_heads=2, BLOCK_T=32, BLOCK_H=4, dead_id=-1,
        lookback_depth=2, lookback_row_stride=2, lookback_col_stride=1,
        lookback_mask_row_stride=2, lookback_mask_col_stride=1,
    )
    torch.npu.synchronize()
    print("OK")

if __name__ == "__main__":
    main()
```

Observed on our machines (stable, deterministic):

```
triton.compiler.errors.MLIRCompilationError:
[ConvertLinalgRToBinary] encounters error:
LLVM ERROR: The buffer memory has been released and cannot be used again!
```

Replacing the entire `while` search with `req = tl.full((BLOCK_T,), 0,
tl.int32).to(tl.int64)` (keeping everything else) makes the same kernel compile and run
correctly; replacing only the `if shift > 0:` broadcast store with an unreachable guard
(`if shift > 999:`) also works. Removing `%` from the broadcast only (keeping the 2D add)
does **not** make it pass, so `%` is not the trigger.

## Suggested investigation

Probe `ConvertLinalgRToBinary`'s handling of (a) loop-carried int64 vectors feeding
(a) a long sequence of masked strided loads and (b) a final 2D broadcast store. The
`"-The buffer memory has been released and cannot be used again!-"` message hints at a
live-range/allocation bug: buffers for values in one region are being released again
when the lowering re-visits them in a later region.

## Environment

- Hardware: Ascend 910 (A3 series)
- CANN: 9.1.0; Python: 3.12.13
- torch: 2.10.0+cpu; torch_npu: 2.10.0.post4
- triton (imported at runtime): 3.2.0; triton-ascend: 3.2.2

**Comment by hongziqi** [comment]
Thx your feedback. We reproduced the failure on the **official Triton-Ascend 3.2.2 release** package.
We also re-verified the original kernel on that 3.2.2 daily package and on the 3.6 daily package: compilation and execution succeed (`OK`)., suggest use daliy package.

<img width="1026" height="666" alt="Image" src="https://github.com/user-attachments/assets/407cf559-9e0e-4b33-bf84-f69202fbaa8b" />
